### PR TITLE
[Lang] Add parameter 'order' to specify layout for scalar, vector, matrix fields

### PR DESCRIPTION
**Pull request by strongoier**
Related issue = #5491

This PR adds support for the following syntax as described in #5491:
```python
a = ti.field(ti.i32, shape=(128, 64, 32), order='kij')
b = ti.Vector.field(3, ti.i32, shape=(128, 64), order='ji', layout=ti.Layout.SOA)
```

### CODE REVIEW

File: python/taichi/lang/matrix.py [hunk 1]
```diff
@@ -1136,43 +1138,68 @@ def field(cls,
         impl.get_runtime().matrix_fields.append(entries)
 
         if shape is None:
-            assert offset is None, "shape cannot be None when offset is being set"
-
-        if shape is not None:
+            if offset is not None:
+                raise TaichiSyntaxError(
+                    'Shape cannot be None when offset is set')
```
Comment (k-ye) [inline python/taichi/lang/matrix.py:1143]: Super nit: `shape`, `offset` are param names, so let's keep it all lowercases.

**Review by k-ye — APPROVED** [review]
LGTM!

File: python/taichi/lang/matrix.py (hunk 1 above)
Comment (strongoier) [reply python/taichi/lang/matrix.py]: Fixed

**Review by yuanming-hu — APPROVED** [review]
Amazing job! LGTM!

**Pull request merged**